File: DID.Swift.SDK/ElastosDIDSDKTests/DummyAdapter.swift
import Foundation
@testable import ElastosDIDSDK
import SPVWrapper

class DummyAdapter: DIDAdapter, DIDResolver {
    
    private var verbose: Bool = false
    private var idtxs: Array<IDTransactionInfo> = []
    
    public init(_ verbose: Bool) {
        self.verbose = verbose
    }

    public init() {
        self.verbose = false
    }

    private static func generateTxId() -> String {
        var str = ""
        while(str.count < 32){
            let re = randomCustom(min: 0, max: 9)
            let r = decTohex(number: re)
            str = "\(str)\(r)"
        }

        return str
    }
    
    func createIdTransaction(_ payload: String, _ memo: String?, _ confirms: Int, _ callback: @escaping (String, Int, String?) -> Void) {
        
        do {
            let txid = try createIdTransaction(payload, memo)
            callback(txid, 0, nil);
        }
        catch {
            // TODO: callback("")
            callback("", -1, error.localizedDescription);
        }
    }
    
    func createIdTransaction(_ payload: String, _ memo: String?) throws -> String {
        
        do {
            let request: IDChainRequest = try IDChainRequest.fromJson(payload)
            
            if (verbose) {
                print("ID Transaction: \(request.operation)\(request.did!)")
                print("     \(request.toJson(false))\(request.did!)" )
                
                if request.operation != IDChainRequestOperation.DEACTIVATE {
                    print("     \(request.document!.toString(true))" )
                }
            }
            
            if !request.isValid {
                throw TestError.failue("Invalid ID transaction request.")
            }
            if request.operation != IDChainRequestOperation.DEACTIVATE {
                if !request.document!.isValid {
                    throw TestError.failue("Invalid DID Document.")
                }
            }
            var ti = getLastTransaction(request.did!)
            
            switch request.operation {
            case .CREATE: do {
                guard ti == nil else {
                    throw TestError.failue("DID already exist.")
                }
                break
                }
            case .UPDATE: do {
                guard ti != nil else {
                    throw TestError.failue("DID not exist.")
                }
                
                guard ti!.operation != IDChainRequestOperation.DEACTIVATE else {
                    throw TestError.failue("DID already dactivated.")
                }
                
                guard request.previousTransactionId == ti!.transactionId else {
                    throw TestError.failue("Previous transaction id missmatch.")
                }
                break
                }
            case .DEACTIVATE: do {
                guard ti != nil else {
                    throw TestError.failue("DID not exist.")
                }
                
                guard ti!.operation != IDChainRequestOperation.DEACTIVATE else {
                    throw TestError.failue("DID already dactivated.")
                }
                break
                }
            }
            //TODO: Date()
            ti = IDTransactionInfo(DummyAdapter.generateTxId(), Date(), request)
            idtxs.append(ti!)
            
            return ti!.transactionId
        } catch {
            print(error)
            throw error
        }
    }

    func resolve(_ requestId: String, _ did: String, _ all: Bool) throws -> Data {
        
        if (verbose) {
            print("Resolve: " + did + "...")
        }
        
        var d = did
        let hashStr = did.prefix(12)
        
        if hashStr != "did:elastos:" {
            d = "did:elastos:\(did)"
        }
        let target: DID = try DID(d)
        var matched: Bool = false
        let generator = JsonGenerator()
        generator.writeStartObject()
        generator.writeStringField("id", requestId)
        generator.writeStringField("jsonrpc", "2.0")
        generator.writeFieldName("result")
        generator.writeStartObject()
        generator.writeStringField("did", target.toString())

        var status = 3
        let last = getLastTransaction(target)
        if last != nil {
            if last!.operation == IDChainRequestOperation.DEACTIVATE {
                status = 2
            }
            else {
                if last!.request.document!.isExpired {
                    status = 1
                }
                else {
                    status = 0
                }
            }
            matched = true
        }
        generator.writeNumberField("status", status)

        if status != 3 {
            generator.writeFieldName("transaction")
            generator.writeStartArray()
            let reversedArr: Array = idtxs.reversed()
            for ti in reversedArr {
                if ti.did == target {
                    ti.toJson(generator)
                    if (!all) {
                        break
                    }
                }
            }
            generator.writeEndArray()
        }
        generator.writeEndObject()
        generator.writeEndObject()

        if (verbose) {
            print(matched ? "success" : "failed")
        }
        return generator.toString().data(using: .utf8)!
    }
    
    public func reset() {
        idtxs.removeAll()
    }
    
    private func getLastTransaction(_ did: DID) -> IDTransactionInfo? {
        let reversedArr = idtxs.reversed()
        for ti in reversedArr {
            if ti.did == did {
                return ti
            }
        }
        return nil
    }
    
    static func randomCustom(min: Int, max: Int) -> Int {
        //  [min, max)  [0, 100)
        //        var x = arc4random() % UInt32(max);
        //        return Int(x)
        // [min, max）
        let y = arc4random() % UInt32(max) + UInt32(min)
        return Int(y)
    }
    
    static func decTohex(number:Int) -> String {
        return String(format: "%0X", number)
    }
}

class IDTx {
    
    private var txId: String
    private var timestamp: Date
    private var request: IDChainRequest
    
    init(_ request: IDChainRequest) {
        self.txId = IDTx.generateTxId()
        self.timestamp = Date() // TODO:
        self.request = request
    }
    
    public func getDid() -> DID {
        return request.did!
    }
    
    private static func generateTxId() -> String {
        var str = ""
        while(str.count < 32){
            let re = randomCustom(min: 0, max: 9)
            let r = decTohex(number: re)
            str = "\(str)\(r)"
        }

        return str
    }
    
    public func getOperation() -> IDChainRequestOperation {
        return request.operation
    }

    public func getPayload() -> String {
        return request.toJson(false)
    }
    
    static func randomCustom(min: Int, max: Int) -> Int {
            //  [min, max)  [0, 100)
            //        var x = arc4random() % UInt32(max);
            //        return Int(x)
            // [min, max）
            let y = arc4random() % UInt32(max) + UInt32(min)
            return Int(y)
    }
    
    static func decTohex(number:Int) -> String {
         return String(format: "%0X", number)
     }
    
    public func toJson() throws -> String {
        let json = request.toJson(false)
        return "{\"id\": \"1\",\"jsonrpc\":\"2.0\",\"result\":[\(json)]}"
    }
}

File: DID.Java.SDK/src/main/java/org/elastos/did/adapter/DummyAdapter.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did.adapter;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.Calendar;
import java.util.LinkedList;
import java.util.Random;

import org.elastos.did.Constants;
import org.elastos.did.DID;
import org.elastos.did.DIDAdapter;
import org.elastos.did.backend.IDChainRequest;
import org.elastos.did.backend.IDTransactionInfo;
import org.elastos.did.exception.DIDResolveException;
import org.elastos.did.exception.DIDTransactionException;

import com.fasterxml.jackson.core.JsonEncoding;
import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;

public class DummyAdapter implements DIDAdapter {
	private static Random random = new Random();

	private boolean verbose;
	private LinkedList<IDTransactionInfo> idtxs;

	public DummyAdapter(boolean verbose) {
		this.verbose = verbose;

		idtxs = new LinkedList<IDTransactionInfo>();
	}

	public DummyAdapter() {
		this(false);
	}

	private static String generateTxid() {
        StringBuffer sb = new StringBuffer();
        while(sb.length() < 32){
            sb.append(Integer.toHexString(random.nextInt()));
        }

        return sb.toString();
    }

	private IDTransactionInfo getLastTransaction(DID did) {
		for (IDTransactionInfo ti : idtxs) {
			if (ti.getDid().equals(did))
				return ti;
		}

		return null;
	}

	private String createIdTransaction(String payload, String memo)
			throws DIDTransactionException {
		IDChainRequest request = IDChainRequest.fromJson(payload);

		if (verbose) {
			System.out.println("ID Transaction: " + request.getOperation()
					+ "[" + request.getDid() + "]");
			System.out.println("    " + request.toJson(false));

			if (request.getOperation() != IDChainRequest.Operation.DEACTIVATE)
				System.out.println("    " + request.getDocument().toString(true));
		}

		if (!request.isValid())
			throw new DIDTransactionException("Invalid ID transaction request.");

		if (request.getOperation() != IDChainRequest.Operation.DEACTIVATE) {
			if (!request.getDocument().isValid())
				throw new DIDTransactionException("Invalid DID Document.");
		}

		IDTransactionInfo ti = getLastTransaction(request.getDid());

		switch (request.getOperation()) {
		case CREATE:
			if (ti != null)
				throw new DIDTransactionException("DID already exist.");

			break;

		case UPDATE:
			if (ti == null)
				throw new DIDTransactionException("DID not exist.");

			if (ti.getOperation() == IDChainRequest.Operation.DEACTIVATE)
				throw new DIDTransactionException("DID already dactivated.");

			if (!request.getPreviousTxid().equals(ti.getTransactionId()))
				throw new DIDTransactionException("Previous transaction id missmatch.");

			break;

		case DEACTIVATE:
			if (ti == null)
				throw new DIDTransactionException("DID not exist.");

			if (ti.getOperation() == IDChainRequest.Operation.DEACTIVATE)
				throw new DIDTransactionException("DID already dactivated.");

			break;
		}

		ti = new IDTransactionInfo(generateTxid(),
				Calendar.getInstance(Constants.UTC).getTime(), request);
		idtxs.add(0, ti);

		return ti.getTransactionId();
	}

	@Override
	public void createIdTransaction(String payload, String memo,
			int confirms, TransactionCallback callback) {
		try {
			String txid = createIdTransaction(payload, memo);
			callback.accept(txid, 0, null);
		} catch (Exception e) {
			callback.accept(null, -1, e.getMessage());
		}
	}

	@Override
	public InputStream resolve(String requestId, String did, boolean all)
			throws DIDResolveException {
		ByteArrayOutputStream os = new ByteArrayOutputStream(4096);
		JsonFactory factory = new JsonFactory();
		boolean matched = false;

		if (verbose)
			System.out.print("Resolve: " + did + "...");

		if (!did.startsWith("did:elastos:"))
			did = "did:elastos:" + did;

		try {
			JsonGenerator generator = factory.createGenerator(os, JsonEncoding.UTF8);
			generator.writeStartObject();

			generator.writeStringField("id", requestId);
			generator.writeStringField("jsonrpc", "2.0");
			generator.writeFieldName("result");
			generator.writeStartObject();

			DID target = new DID(did);

			generator.writeStringField("did", target.toString());

			int status = 3;
			IDTransactionInfo last = getLastTransaction(target);
			if (last != null) {
				if (last.getOperation() == IDChainRequest.Operation.DEACTIVATE) {
					status = 2;
			    } else {
					if (last.getRequest().getDocument().isExpired())
						status = 1;
					else
						status = 0;
				}

				matched = true;
			}

			generator.writeNumberField("status", status);

			if (status != 3) {
				generator.writeFieldName("transaction");
				generator.writeStartArray();
				for (IDTransactionInfo ti : idtxs) {
					if (ti.getDid().equals(target)) {
						ti.toJson(generator);

						if (!all)
							break;
					}
				}
				generator.writeEndArray();
			}

			generator.writeEndObject();
			generator.writeEndObject();
			generator.close();
		} catch (Exception e) {
			throw new DIDResolveException(e);
		}

		if (verbose)
			System.out.println(matched ? "success" : "failed");

		return new ByteArrayInputStream(os.toByteArray());
	}

	public void reset() {
		idtxs.clear();
	}
}
